fencing/remote.c

```diff
@@ -999,14 +999,27 @@ create_remote_stonith_op(const char *client, xmlNode * request, gboolean peer)
 
     if (op->call_options & st_opt_cs_nodeid) {
         int nodeid = crm_atoi(op->target, NULL);
-        crm_node_t *node = crm_get_peer(nodeid, NULL);
+        crm_node_t *node = crm_find_peer(nodeid, NULL);
+        const char *nodename = NULL;
 
         /* Ensure the conversion only happens once */
         op->call_options &= ~st_opt_cs_nodeid;
 
         if (node && node->uname) {
+            nodename = node->uname;
+
+        } else {
+            const char *last_known_name = g_hash_table_lookup(known_peer_names, GINT_TO_POINTER(nodeid));
+
+            if (last_known_name) {
+                nodename = last_known_name;
+            }
+        }
```

Comment:
Let's functionize this if-else block, since it's used three times.

I like the switch to crm_find_peer(), to avoid re-adding a cache entry that might never get cleaned, but we lose crm_get_peer()'s check for get_node_name(). So, let's check get_node_name() before checking known_peer_names, since it's more accurate if it returns anything. In the particular scenario we're interested in here, it won't, but there may be cases where it does.